### DISCUSSION TITLE
C++: Linking issue when using dynamic libraries

### DISCUSSION THREAD

**Discussion by SonnleitnerF** [foxglove/mcap#1202] [General] [answered]
I am using MCAP within my codebase and have run into a linking issue when I am trying to include the reader/writer within header files. Until now I could prevent such issues by defining the `MCAP_IMPLEMENTATION` macro in the top level CMakeLists.txt and including the mcap reader/writer only in source files. Due to the nature of the code this is no longer possible as I am working with templates. A small excerpt of the linker error can be found below. 

`error LNK2005: "public: __cdecl mcap::MetadataIndex::MetadataIndex(struct mcap::Metadata const &,unsigned __int64)" (??0MetadataIndex@mcap@@QEAA@AEBUMetadata@1@_K@Z) already defined in *********.cpp.obj`

I am using MSVC 17.9.7 with conan and cmake

I could not find any thing in the documentation how to fix this?

**Comment by indirectlylit** [comment]
Hi @SonnleitnerF -

We recently added a link in the docs to a community repo which can help smooth CMake integration: [olympus-robotics/mcap_builder](https://github.com/olympus-robotics/mcap_builder). In the readme, they note:

> Do not add `MCAP_IMPLEMENTATION` macro, as that is handled for you by this wrapper.

Curious if this might help with your use-case?

**Reply by SonnleitnerF** [reply]
Thanks for that. It seems the go to solution is to prebuilt the library and use it in you own code. I was hoping to not have to do that, but thanks you 😃

**Reply by indirectlylit** [reply]
according to the readme, you should also be able to build it as part of your project:

https://github.com/olympus-robotics/mcap_builder?tab=readme-ov-file#build-mcap-locally

does that describe your use-case?

**Reply by SonnleitnerF** [reply]
This did not work unfortunately. I got some errors showing some incompatibility issues on windows with the C++ version i was using (C++20 with the C++23 features MSVC 17.9.7 supports). After that I did not further try this approach

**Reply by indirectlylit** [reply]
(cc @filippobrizzi - fyi)

**Reply by filippobrizzi** [reply]
unfortunately, I don't have a way to test on Windows to try and fix it, but @SonnleitnerF  if you have time to contribute to [olympus-robotics/mcap_builder](https://github.com/olympus-robotics/mcap_builder) that would be awesome for the next person with your same issue.

**Reply by SonnleitnerF** [reply]
I will keep it in mind! Thanks for the infos

**Comment by jtbandes** [comment]
MCAP_IMPLEMENTATION is supposed to be defined in only one translation unit, it should not be defined globally.

**Reply by SonnleitnerF** [reply]
what if I want to use it in different libraries though? Then I get "symbol unknown" linker errors, and if I declare it in every library I want to use it in, I get the mentioned "already defined symbol" linker error.

**Comment by SonnleitnerF** [comment, marked as answer]
I now resolved this issue by prebuilding mcap and linking against it. This way the macro is not needed anymore and no linker issues occur. 

I leave the discussion open for other potential solutions, if other developers experiencing similar issues can have different approaches
